src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.Core.cs

```diff
@@ -18,58 +18,60 @@
     {
-        /// <summary>
-        /// Once the assembly path is chosen, load it directly from disk at that location
-        /// </summary>
-        LoadFromDisk,
+        internal static IAnalyzerAssemblyResolver DiskAnalyzerAssemblyResolver => DiskResolver.Instance;
+        internal static IAnalyzerAssemblyResolver StreamAnalyzerAssemblyResolver => StreamResolver.Instance;
 
-        /// <summary>
-        /// Once the assembly path is chosen, read the contents of disk and load from memory
-        /// </summary>
-        /// <remarks>
-        /// While Windows supports this option it comes with a significant performance penalty due
-        /// to anti virus scans. It can have a load time of 300-500ms while loading from disk 
-        /// is generally 1-2ms. Use this with caution on Windows.
-        /// </remarks>
-        LoadFromStream
-    }
-
-    internal partial class AnalyzerAssemblyLoader
-    {
-        private readonly AssemblyLoadContext _compilerLoadContext;
         private readonly Dictionary<string, DirectoryLoadContext> _loadContextByDirectory = new Dictionary<string, DirectoryLoadContext>(StringComparer.Ordinal);
-        private readonly AnalyzerLoadOption _loadOption;
 
-        internal AssemblyLoadContext CompilerLoadContext => _compilerLoadContext;
-        internal AnalyzerLoadOption AnalyzerLoadOption => _loadOption;
+        public IAnalyzerAssemblyResolver CompilerAnalyzerAssemblyResolver { get; }
+        public AssemblyLoadContext CompilerLoadContext { get; }
+        public ImmutableArray<IAnalyzerAssemblyResolver> AnalyzerAssemblyResolvers { get; }
 
-        internal AnalyzerAssemblyLoader(ImmutableArray<IAnalyzerAssemblyResolver> externalResolvers)
-            : this(null, AnalyzerLoadOption.LoadFromDisk, externalResolvers)
+        internal AnalyzerAssemblyLoader()
+            : this([], [DiskAnalyzerAssemblyResolver], null)
         {
         }
 
-        internal AnalyzerAssemblyLoader(AssemblyLoadContext? compilerLoadContext, AnalyzerLoadOption loadOption, ImmutableArray<IAnalyzerAssemblyResolver> externalResolvers)
+        /// <summary>
+        /// Create a new <see cref="AnalyzerAssemblyLoader"/> with the given resolvers.
+        /// </summary>
+        /// <param name="pathResolvers"></param>
+        /// <param name="assemblyResolvers"></param>
+        /// <param name="compilerLoadContext">This is the <see cref="AssemblyLoadContext"/> where the compiler resides. This parameter
+        /// is primarily used for testing purposes but is also useful in hosted scenarios where the compiler may be loaded outside
+        /// the default context.</param>
+        /// <exception cref="ArgumentException"></exception>
+        internal AnalyzerAssemblyLoader(
+            ImmutableArray<IAnalyzerPathResolver> pathResolvers,
+            ImmutableArray<IAnalyzerAssemblyResolver> assemblyResolvers = default,
+            AssemblyLoadContext? compilerLoadContext = null)
```

Comment:
consider non-optional.  in this case i'm on the fence.  if, practically, no clients need to set it, then optional is fine.  but the point about "useful in hosted scenarios" make it feel more like a case where it could be good to be required and to have clients be explicit about what they want here.

Reply:
The use of optional values here is done to keep the rest of the code base sane. The last two parameters here are `#if NET` only. When I make them non-optional then it forces a large number of `#if NET #else` into the IDE, workspaces, etc ... layer. Tried to make that work and it was just ugly. Put the optionals in here to make the construction syntax, in the 99% case, the same between the two and hence make our code more readable. 

I also considered having separate ctors here and making it non-optional in the other. If that is preferred I'm fine with it.

Reply:
that's a perfectly good reason to have this parameter be optional.  keep in that way :) 

#resolved.